File: hw05/bulls/lib/bulls/game.ex
defmodule Bulls.Game do
  def new do
    %{
      secret: get_number(""),
      guesses: MapSet.new(),
      text: "",
      message: "Enter a unique 4-digit number",
      hints: []
    }
  end

  def guess(st, num) do
    cond do
      valid_guess(num) ->
        %{st | guesses: MapSet.put(st.guesses, num)}

      true ->
        %{st | guesses: st.guesses}
    end
  end

  def view(st) do
    num = st.secret
    hints = Enum.map(st.guesses, fn x -> get_hints(num, x) end)

    %{
      hints: hints,
    }
  end

  def valid_guess(guess) do
    guess |> Integer.parse() != :error &&
      guess |> Integer.parse() |> elem(1) == "" &&
      guess |> String.length() == 4 &&
      !(guess |> not_unique())
  end

  def not_unique(text) do
    nums = String.split(text, "", trim: true)
    sorted_nums = Enum.sort(nums)

    sorted_nums |> Enum.at(0) === sorted_nums |> Enum.at(1) ||
      sorted_nums |> Enum.at(1) === sorted_nums |> Enum.at(2) ||
      sorted_nums |> Enum.at(2) === sorted_nums |> Enum.at(3)
  end

  def get_message(text) do
    cond do
      text |> Integer.parse() == :error ->
        "Error: Must enter a 4-digit number"

      text |> Integer.parse() |> elem(1) != "" ->
        "Error: Must enter a 4-digit number"

      text |> String.length() != 4 ->
        "Error: Number is not 4 digits"

      text |> not_unique() ->
        "Error: All 4 digits must be unique"

      true ->
        "Enter a unique 4-digit number"
    end
  end

  def get_number(num) do
    digit = Enum.random(0..9) |> Integer.to_string()

    cond do
      num |> String.length() == 4 ->
        num

      String.contains?(num, digit) ->
        get_number(num)

      true ->
        get_number(num <> digit)
    end
  end

  def get_hints(secret, guess) do
    bulls = get_bulls(secret, guess) |> Integer.to_string()
    cows = get_cows(secret, guess) |> Integer.to_string()
    "" <> guess <> " | Bulls: " <> bulls <> ", Cows: " <> cows
  end

  def get_bulls(secret, guess) do
    len = guess |> String.length()
    secret_tail = secret |> String.slice(1, len)
    guess_tail = guess |> String.slice(1, len)

    cond do
      guess |> String.length() == 0 ->
        0

      String.first(secret) === String.first(guess) ->
        1 + get_bulls(secret_tail, guess_tail)

      true ->
        get_bulls(secret_tail, guess_tail)
    end
  end

  def get_cows(secret, guess) do
    len = guess |> String.length()
    guess_tail = guess |> String.slice(1, len)

    cond do
      guess |> String.length() == 0 ->
        0

      secret |> String.contains?(guess |> String.first()) ->
        if len == 4 do
          1 + get_cows(secret, guess_tail) - get_bulls(secret, guess)
        else
          1 + get_cows(secret, guess_tail)
        end

      true ->
        if len == 4 do
          get_cows(secret, guess_tail) - get_bulls(secret, guess)
        else
          get_cows(secret, guess_tail)
        end
    end
  end
end

File: hw05/bulls/assets/js/game.js
/**
 * I used Prof. Tuck's Hangman lecture code as a starting point.
 * I left some of his code here, such as uniq() and lives_left(), but I had
 * to write a lot of new code to program the necessary logic for Bulls and Cows.
 */
export function uniq(xs) {
    return Array.from(new Set(xs));
}

function not_unique(text) {
    let nums = Array.from(text);
    let sorted_nums = nums.sort();
    return sorted_nums[0] === sorted_nums[1] ||
        sorted_nums[1] === sorted_nums[2] ||
        sorted_nums[2] === sorted_nums[3];
}

export function get_message(text) {
    let message = "";
    if (isNaN(parseInt(text))) {
        message = "Error: Must enter a 4-digit number";
    } else if (text.length < 4) {
        message = "Error: Number is less than 4 digits";
    } else if (not_unique(text)) {
        message = "Error: All 4 digits must be unique";
    } else {
        message = "Enter a unique 4-digit number";
    }
    return message;
}

export function get_number() {
    let solution = "";
    while (solution.length < 4) {
        let digit = Math.floor((Math.random() * 10));
        while (!solution.includes(digit.toString())) {
            solution += digit.toString();
        }
    }
    return solution;
}

export function get_bulls(secret, guess) {
    let bulls = 0;
    for (let i = 0; i < 4; i++) {
        if (guess.charAt(i) === secret.charAt(i)) {
            bulls++;
        }
    }
    return bulls;
}

export function get_cows(secret, guess) {
    let cows = 0;
    for (let i = 0; i < 4; i++) {
        if (secret.includes(guess.charAt(i))) {
            cows++;
        }
    }
    cows -= get_bulls(secret, guess);
    return cows;
}

export function bad_guesses(secret, guesses) {
    let bads = [];
    for (let gg in guesses) {
        let curr = guesses[gg];
        if (parseInt(secret) !== parseInt(curr)) {
            if (!isNaN(parseInt(curr)) && curr.length === 4 && !not_unique(curr)) {
                bads.push(guesses[gg]);
            }
        }
    }
    return uniq(bads);
}

export function lives_left(secret, guesses) {
    return 8 - bad_guesses(secret, guesses).length;
}
